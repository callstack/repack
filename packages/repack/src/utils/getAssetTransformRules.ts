import type { RuleSetRule } from '@rspack/core';
import type { AssetLoaderRemoteOptions } from '../loaders/assetsLoader/options.js';
import {
  ASSET_EXTENSIONS,
  getAssetExtensionsRegExp,
} from './assetExtensions.js';

function getSvgRule(type: 'svgr' | 'xml' | 'uri'): RuleSetRule {
  if (type === 'svgr') {
    return {
      test: /\.svg$/,
      use: [{ loader: '@svgr/webpack', options: { native: true } }],
    };
  }

  return {
    test: /\.svg$/,
    type: type === 'xml' ? 'asset/source' : 'asset/inline',
  };
}

interface GetAssetTransformRulesOptions {
  inline?: boolean;
  remote?: Omit<AssetLoaderRemoteOptions, 'enabled'>;
  svg?: 'svgr' | 'xml' | 'uri';
}

export function getAssetTransformRules({
  inline,
  remote,
  svg,
}: GetAssetTransformRulesOptions = {}) {
  const extensions = svg
    ? ASSET_EXTENSIONS.filter((ext) => ext !== '.svg')
    : ASSET_EXTENSIONS;

  const remoteOptions = remote
    ? {
        enabled: true,
        ...remote,
      }
    : undefined;

  const rules: RuleSetRule[] = [
    {
      test: getAssetExtensionsRegExp(extensions),
      use: {
        loader: '@callstack/repack/assets-loader',
        options: { inline, remote: remoteOptions },
      },
    },
  ];

  if (svg) {
    rules.push(getSvgRule(svg));
  }

  return rules;
}
