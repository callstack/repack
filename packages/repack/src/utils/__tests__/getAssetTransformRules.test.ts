import { describe, expect, it } from 'vitest';
import { getAssetTransformRules } from '../getAssetTransformRules.js';

describe('getAssetTransformRules', () => {
  it('should return default asset transform rules when no options provided', () => {
    const rules = getAssetTransformRules();
    expect(rules).toMatchSnapshot();
  });

  it('should return rules with inline option when provided', () => {
    const rules = getAssetTransformRules({ inline: true });

    // @ts-ignore
    expect(rules[0]?.use?.options?.inline).toEqual(true);
    expect(rules).toMatchSnapshot();
  });

  it('should return rules with remote options when provided', () => {
    const remoteOptions = { publicPath: 'https://example.com/assets' };
    const rules = getAssetTransformRules({ remote: remoteOptions });

    // @ts-ignore
    expect(rules[0]?.use?.options?.remote).toHaveProperty('enabled', true);
    expect(rules).toMatchSnapshot();
  });

  it('should add SVGR rule when svg="svgr"', () => {
    const rules = getAssetTransformRules({ svg: 'svgr' });

    expect(rules).toHaveLength(2);
    expect(rules[1]?.use?.loader).toEqual('@svgr/webpack');
    expect(rules).toMatchSnapshot();
  });

  it('should add XML rule when svg="xml"', () => {
    const rules = getAssetTransformRules({ svg: 'xml' });

    expect(rules).toHaveLength(2);
    // @ts-ignore
    expect(rules[1]?.type).toEqual('asset/source');
    expect(rules).toMatchSnapshot();
  });

  it('should add URI rule when svg="uri"', () => {
    const rules = getAssetTransformRules({ svg: 'uri' });

    expect(rules).toHaveLength(2);
    // @ts-ignore
    expect(rules[1]?.type).toEqual('asset/inline');
    expect(rules).toMatchSnapshot();
  });

  it('should exclude .svg from main asset extensions when svg option is provided', () => {
    const rules = getAssetTransformRules({ svg: 'uri' });
    const ruleTest = rules[0]?.test;
    expect(ruleTest.test('test.svg')).toEqual(false);
  });
});
