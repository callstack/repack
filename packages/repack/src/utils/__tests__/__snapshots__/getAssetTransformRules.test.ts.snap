// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAssetTransformRules > should add SVGR rule when svg="svgr" 1`] = `
[
  {
    "test": /\\\\\\.\\(bmp\\|gif\\|jpg\\|jpeg\\|png\\|psd\\|webp\\|tiff\\|m4v\\|mov\\|mp4\\|mpeg\\|mpg\\|webm\\|aac\\|aiff\\|caf\\|m4a\\|mp3\\|wav\\|html\\|pdf\\|yaml\\|yml\\|otf\\|ttf\\|zip\\|obj\\)\\$/,
    "use": {
      "loader": "@callstack/repack/assets-loader",
      "options": {
        "inline": undefined,
        "remote": undefined,
      },
    },
  },
  {
    "test": /\\\\\\.svg\\$/,
    "use": {
      "loader": "@svgr/webpack",
      "options": {
        "native": true,
      },
    },
  },
]
`;

exports[`getAssetTransformRules > should add URI rule when svg="uri" 1`] = `
[
  {
    "test": /\\\\\\.\\(bmp\\|gif\\|jpg\\|jpeg\\|png\\|psd\\|webp\\|tiff\\|m4v\\|mov\\|mp4\\|mpeg\\|mpg\\|webm\\|aac\\|aiff\\|caf\\|m4a\\|mp3\\|wav\\|html\\|pdf\\|yaml\\|yml\\|otf\\|ttf\\|zip\\|obj\\)\\$/,
    "use": {
      "loader": "@callstack/repack/assets-loader",
      "options": {
        "inline": undefined,
        "remote": undefined,
      },
    },
  },
  {
    "test": /\\\\\\.svg\\$/,
    "type": "asset/inline",
  },
]
`;

exports[`getAssetTransformRules > should add XML rule when svg="xml" 1`] = `
[
  {
    "test": /\\\\\\.\\(bmp\\|gif\\|jpg\\|jpeg\\|png\\|psd\\|webp\\|tiff\\|m4v\\|mov\\|mp4\\|mpeg\\|mpg\\|webm\\|aac\\|aiff\\|caf\\|m4a\\|mp3\\|wav\\|html\\|pdf\\|yaml\\|yml\\|otf\\|ttf\\|zip\\|obj\\)\\$/,
    "use": {
      "loader": "@callstack/repack/assets-loader",
      "options": {
        "inline": undefined,
        "remote": undefined,
      },
    },
  },
  {
    "test": /\\\\\\.svg\\$/,
    "type": "asset/source",
  },
]
`;

exports[`getAssetTransformRules > should return default asset transform rules when no options provided 1`] = `
[
  {
    "test": /\\\\\\.\\(bmp\\|gif\\|jpg\\|jpeg\\|png\\|psd\\|svg\\|webp\\|tiff\\|m4v\\|mov\\|mp4\\|mpeg\\|mpg\\|webm\\|aac\\|aiff\\|caf\\|m4a\\|mp3\\|wav\\|html\\|pdf\\|yaml\\|yml\\|otf\\|ttf\\|zip\\|obj\\)\\$/,
    "use": {
      "loader": "@callstack/repack/assets-loader",
      "options": {
        "inline": undefined,
        "remote": undefined,
      },
    },
  },
]
`;

exports[`getAssetTransformRules > should return rules with inline option when provided 1`] = `
[
  {
    "test": /\\\\\\.\\(bmp\\|gif\\|jpg\\|jpeg\\|png\\|psd\\|svg\\|webp\\|tiff\\|m4v\\|mov\\|mp4\\|mpeg\\|mpg\\|webm\\|aac\\|aiff\\|caf\\|m4a\\|mp3\\|wav\\|html\\|pdf\\|yaml\\|yml\\|otf\\|ttf\\|zip\\|obj\\)\\$/,
    "use": {
      "loader": "@callstack/repack/assets-loader",
      "options": {
        "inline": true,
        "remote": undefined,
      },
    },
  },
]
`;

exports[`getAssetTransformRules > should return rules with remote options when provided 1`] = `
[
  {
    "test": /\\\\\\.\\(bmp\\|gif\\|jpg\\|jpeg\\|png\\|psd\\|svg\\|webp\\|tiff\\|m4v\\|mov\\|mp4\\|mpeg\\|mpg\\|webm\\|aac\\|aiff\\|caf\\|m4a\\|mp3\\|wav\\|html\\|pdf\\|yaml\\|yml\\|otf\\|ttf\\|zip\\|obj\\)\\$/,
    "use": {
      "loader": "@callstack/repack/assets-loader",
      "options": {
        "inline": undefined,
        "remote": {
          "enabled": true,
          "publicPath": "https://example.com/assets",
        },
      },
    },
  },
]
`;
